File: python/krpc/service.py
import re
import keyword
from collections import defaultdict
from krpc.attributes import Attributes
from krpc.types import Types, DynamicType, DefaultArgument
from krpc.decoder import Decoder

_regex_multi_uppercase = re.compile(r'([A-Z]+)([A-Z][a-z0-9])')
_regex_single_uppercase = re.compile(r'([a-z0-9])([A-Z])')
_regex_underscores = re.compile(r'(.)_')

def _to_snake_case(camel_case):
    """ Convert camel case to snake case, e.g. GetServices -> get_services """
    result = re.sub(_regex_underscores, r'\1__', camel_case)
    result = re.sub(_regex_single_uppercase, r'\1_\2', result)
    return re.sub(_regex_multi_uppercase, r'\1_\2', result).lower()

def _signature(param_types, return_type):
    """ Generate a signature for a procedure that can be used as its docstring """
    if len(param_types) == 0 and return_type == None:
        return ''
    types = [x.python_type.__name__ for x in param_types]
    sig = ','.join(types)
    if len(types) == 0:
        sig = '()'
    elif len(types) > 1:
        sig = '('+sig+')'
    if return_type is not None:
        sig += ' -> '+return_type.python_type.__name__
    return sig

def _as_literal(value, typ):
    if typ.python_type == str:
        return '\''+value+'\''
    return str(value)

def _update_param_names(names):
    """ Given a list of parameter names, append underscores to reserved keywords
        without causing parameter names to clash """
    newnames = []
    for name in names:
        if keyword.iskeyword(name):
            name = name+'_'
            while name in names:
                name += '_'
        newnames.append(name)
    return newnames

def _construct_func(invoke, service_name, procedure_name, prefix_param_names, param_names,
                    param_types, param_required, param_default, return_type):
    """ Build function to invoke a remote procedure """

    prefix_param_names = _update_param_names(prefix_param_names)
    param_names = _update_param_names(param_names)

    params = []
    for name,required,default,typ in zip(param_names, param_required, param_default, param_types):
        if not required:
            name += ' = DefaultArgument('+repr(_as_literal(default,typ))+')'
        params.append(name)

    invoke_args = [
        '\''+str(service_name)+'\'',
        '\''+str(procedure_name)+'\'',
        '['+','.join(param_names)+']',
        '{}',
        'param_names',
        'param_types',
        'return_type'
    ]
    code = 'lambda ' + ', '.join(prefix_param_names + params) + ': invoke('+', '.join(invoke_args)+')'
    context = {
        'invoke': invoke,
        'DefaultArgument': DefaultArgument,
        'param_names': param_names,
        'param_types': param_types,
        'return_type': return_type
    }
    return eval(code, context)

def create_service(client, service):
    """ Create a new service type """
    cls = type(str(service.name), (ServiceBase,), {'_client': client, '_name': service.name})

    # Add class types to service
    for cls2 in service.classes:
        cls._add_service_class(cls2)

    # Add enumeration types to service
    for enum in service.enumerations:
        cls._add_service_enumeration(enum)

    # Add procedures
    for procedure in service.procedures:
        if Attributes.is_a_procedure(procedure.attributes):
            cls._add_service_procedure(procedure)

    # Add properties
    properties = defaultdict(lambda: [None,None])
    for procedure in service.procedures:
        if Attributes.is_a_property_accessor(procedure.attributes):
            name = Attributes.get_property_name(procedure.attributes)
            if Attributes.is_a_property_getter(procedure.attributes):
                properties[name][0] = procedure
            else:
                properties[name][1] = procedure
    for name, procedures in properties.items():
        cls._add_service_property(name, procedures[0], procedures[1])

    # Add class methods
    for procedure in service.procedures:
        if Attributes.is_a_class_method(procedure.attributes):
            class_name = Attributes.get_class_name(procedure.attributes)
            method_name = Attributes.get_class_method_name(procedure.attributes)
            cls._add_service_class_method(class_name, method_name, procedure)

    # Add static class methods
    for procedure in service.procedures:
        if Attributes.is_a_class_static_method(procedure.attributes):
            class_name = Attributes.get_class_name(procedure.attributes)
            method_name = Attributes.get_class_method_name(procedure.attributes)
            cls._add_service_class_static_method(class_name, method_name, procedure)

    # Add class properties
    properties = defaultdict(lambda: [None,None])
    for procedure in service.procedures:
        if Attributes.is_a_class_property_accessor(procedure.attributes):
            class_name = Attributes.get_class_name(procedure.attributes)
            property_name = Attributes.get_class_property_name(procedure.attributes)
            key = (class_name, property_name)
            if Attributes.is_a_class_property_getter(procedure.attributes):
                properties[key][0] = procedure
            else:
                properties[key][1] = procedure
    for (class_name, property_name), procedures in properties.items():
        cls._add_service_class_property(class_name, property_name, procedures[0], procedures[1])

    return cls()

class ServiceBase(DynamicType):
    """ Base class for service objects, created at runtime using information received from the server. """

    @classmethod
    def _add_service_class(cls, remote_cls):
        """ Add a class type """
        name = remote_cls.name
        class_type = cls._client._types.as_type('Class(' + cls._name + '.' + name + ')')
        setattr(cls, name, class_type.python_type)

    @classmethod
    def _add_service_enumeration(cls, enum):
        """ Add an enumeration type """
        name = enum.name
        enum_type = cls._client._types.as_type('Enum(' + cls._name + '.' + name + ')')
        enum_type.set_values(dict((str(_to_snake_case(x.name)), x.value) for x in enum.values))
        setattr(cls, name, enum_type.python_type)

    @classmethod
    def _parse_procedure(cls, procedure):
        param_names = [_to_snake_case(param.name) for param in procedure.parameters]
        param_types = [cls._client._types.get_parameter_type(i, param.type, procedure.attributes) for i,param in enumerate(procedure.parameters)]
        param_required = [not param.HasField('default_argument') for param in procedure.parameters]
        param_default = []
        for param,typ in zip(procedure.parameters, param_types):
            if param.HasField('default_argument'):
                param_default.append(Decoder.decode(param.default_argument, typ))
            else:
                param_default.append(None)
        return_type = None
        if procedure.HasField('return_type'):
            return_type = cls._client._types.get_return_type(procedure.return_type, procedure.attributes)
        return param_names, param_types, param_required, param_default, return_type

    @classmethod
    def _add_service_procedure(cls, procedure):
        """ Add a procedure """
        param_names, param_types, param_required, param_default, return_type = cls._parse_procedure(procedure)
        func = _construct_func(cls._client._invoke, cls._name, procedure.name, [], param_names, param_types, param_required, param_default, return_type)
        build_request = _construct_func(cls._client._build_request, cls._name, procedure.name, [], param_names, param_types, param_required, param_default, return_type)
        setattr(func, '_build_request', build_request)
        setattr(func, '_return_type', return_type)
        name = str(_to_snake_case(procedure.name))
        return cls._add_static_method(name, func, doc=None)

    @classmethod
    def _add_service_property(cls, name, getter=None, setter=None):
        """ Add a property """
        if getter:
            getter_name = getter.name
            _,_,_,_,return_type = cls._parse_procedure(getter)
            getter = _construct_func(cls._client._invoke, cls._name, getter_name, ['self'], [], [], [], [], return_type)
            build_request = _construct_func(cls._client._build_request, cls._name, getter_name, ['self'], [], [], [], [], return_type)
            setattr(getter, '_build_request', build_request)
            setattr(getter, '_return_type', return_type)
        if setter:
            param_names, param_types, _,_,_ = cls._parse_procedure(setter)
            setter = _construct_func(cls._client._invoke, cls._name, setter.name, ['self'], param_names, param_types, [True], [None], None)
        name = str(_to_snake_case(name))
        return cls._add_property(name, getter, setter, doc=None)

    @classmethod
    def _add_service_class_method(cls, class_name, method_name, procedure):
        """ Add a method to a class """
        class_cls = cls._client._types.as_type('Class('+cls._name+'.'+class_name+')').python_type
        param_names, param_types, param_required, param_default, return_type = cls._parse_procedure(procedure)
        # Rename this to self if it doesn't cause a name clash
        if 'self' not in param_names:
            param_names[0] = 'self'
        func = _construct_func(cls._client._invoke, cls._name, procedure.name, [], param_names, param_types, param_required, param_default, return_type)
        build_request = _construct_func(cls._client._build_request, cls._name, procedure.name, [], param_names, param_types, param_required, param_default, return_type)
        setattr(func, '_build_request', build_request)
        setattr(func, '_return_type', return_type)
        name = str(_to_snake_case(method_name))
        class_cls._add_method(name, func, doc=None)

    @classmethod
    def _add_service_class_static_method(cls, class_name, method_name, procedure):
        """ Add a static method to a class """
        class_cls = cls._client._types.as_type('Class('+cls._name+'.'+class_name+')').python_type
        param_names, param_types, param_required, param_default, return_type = cls._parse_procedure(procedure)
        func = _construct_func(cls._client._invoke, cls._name, procedure.name, [], param_names, param_types, param_required, param_default, return_type)
        build_request = _construct_func(cls._client._build_request, cls._name, procedure.name, [], param_names, param_types, param_required, param_default, return_type)
        setattr(func, '_build_request', build_request)
        setattr(func, '_return_type', return_type)
        name = str(_to_snake_case(method_name))
        class_cls._add_static_method(name, func, doc=None)

    @classmethod
    def _add_service_class_property(cls, class_name, property_name, getter=None, setter=None):
        """ Add a property to a class """
        class_cls = cls._client._types.as_type('Class('+cls._name+'.'+class_name+')').python_type
        if getter:
            getter_name = getter.name
            param_names, param_types, param_required, param_default, return_type = cls._parse_procedure(getter)
            # Rename this to self if it doesn't cause a name clash
            if 'self' not in param_names:
                param_names[0] = 'self'
            getter = _construct_func(cls._client._invoke, cls._name, getter_name, [], param_names, param_types, [True], [None], return_type)
            build_request = _construct_func(cls._client._build_request, cls._name, getter_name, [], param_names, param_types, [True], [None], return_type)
            setattr(getter, '_build_request', build_request)
            setattr(getter, '_return_type', return_type)
        if setter:
            param_names, param_types, param_required, param_default, return_type = cls._parse_procedure(setter)
            setter = _construct_func(cls._client._invoke, cls._name, setter.name, [], param_names, param_types, [True,True], [None,None], None)
        property_name = str(_to_snake_case(property_name))
        return class_cls._add_property(property_name, getter, setter, doc=None)

File: lua/krpc/service.lua
local class = require 'pl.class'
local seq = require 'pl.seq'
local stringx = require 'pl.stringx'
local bind1 = require 'pl.func'.bind1
local List = require 'pl.List'
local Map = require 'pl.Map'
local Types = require 'krpc.types'
local Attributes = require 'krpc.attributes'
local decoder = require 'krpc.decoder'

local service = {}

local regex_multi_uppercase = '([A-Z]+)([A-Z][a-z0-9])'
local regex_single_uppercase = '([a-z0-9])([A-Z])'
local regex_underscores = '(.)_'

--- Convert camel case to snake case, e.g. GetServices -> get_services
function service.to_snake_case(camel_case)
  local result = camel_case:gsub(regex_underscores, '%1__')
  result = result:gsub(regex_single_uppercase, '%1_%2')
  return result:gsub(regex_multi_uppercase, '%1_%2'):lower()
end

local function get_names(xs)
  return seq.copy(seq.map(function (x) return service.to_snake_case(x.name) end, xs))
end

local function get_types(types, xs, attrs)
  local result = List{}
  for i,x in ipairs(xs) do
    result:append(types:get_parameter_type(i, x.type, List(attrs)))
  end
  return result
end

local ServiceBase = class(Types.DynamicType)

function ServiceBase:_parse_procedure(procedure)
  local param_names = get_names(procedure.parameters)
  local param_types = get_types(self._client._types, procedure.parameters, procedure.attributes)
  local param_required = seq.copy(seq.map(function (x) return not x:HasField('default_argument') end, procedure.parameters))
  local param_default = List{}
  for param,typ in seq.zip(procedure.parameters, param_types) do
    if param:HasField('default_argument') then
      param_default:append(decoder.decode(param.default_argument, typ))
    else
      param_default:append(nil)
    end
  end
  local return_type = nil
  if procedure:HasField('return_type') then
    return_type = self._client._types:get_return_type(procedure.return_type, List(procedure.attributes))
  end
  return param_names, param_types, param_required, param_default, return_type
end

local function _construct_func(invoke, service_name, procedure_name, prefix_param_names, param_names,
                               param_types, param_required, param_default, return_type)
  prefix_param_names = prefix_param_names or List{}
  param_names = param_names or List{}
  local body =
    'return invoke(' ..
    stringx.join(
      ',',
      {'service_name',
       'procedure_name',
       'nil',
       'Map{'..stringx.join(',', seq.copy(seq.map(function (x) return x..'='..x end, param_names)))..'}',
       'param_names',
       'param_types',
       'return_type'
      }) ..
    ')'
  local func = 'return function (' .. stringx.join(',', prefix_param_names..param_names) .. ') ' .. body .. ' end'
  local wrapper =
    'return function (invoke,service_name,procedure_name,param_names,param_types,return_type,Map) ' .. func .. ' end'
  local callable = assert(loadstring(wrapper, '_construct_func('..service_name..','..procedure_name..')'))()
  return callable(invoke,service_name,procedure_name,param_names,param_types,return_type,Map)

  --return function (...)
  --  local skip = prefix_param_names:len() -- ignore prefix parameters
  --  local args = List(table.pack(...)):slice(skip+1)
  --  print(args)
  --  return invoke(service_name, procedure_name, args, nil, param_names, param_types, return_type)
  --end
end

function ServiceBase:_add_service_class(cls)
  local class_type = self._client._types:as_type('Class(' .. self._name .. '.' .. cls.name .. ')')
  self[cls.name] = class_type.lua_type
end

function ServiceBase:_add_service_procedure(procedure)
  local param_names, param_types, param_required, param_default, return_type = self:_parse_procedure(procedure)
  local func = _construct_func(self._invoke, self._name, procedure.name, List{}, param_names, param_types, param_required, param_default, return_type)
  --build_request = ...
  name = service.to_snake_case(procedure.name)
  return self:_add_static_method(name, func)
end

function ServiceBase:_add_service_property(name, getter, setter)
  if getter then
    getter_name = getter.name
    _,_,_,_,return_type = self:_parse_procedure(getter)
    getter = _construct_func(self._invoke, self._name, getter_name, List{'self'}, nil, nil, nil, nil, return_type)
  end
  if setter then
    setter_name = setter.name
    param_names, param_types, _,_,_ = self:_parse_procedure(setter)
    setter = _construct_func(self._invoke, self._name, setter_name, List{'self'}, param_names, param_types, List{true}, List{nil}, nil)
  end
  name = service.to_snake_case(name)
  return self:_add_property(name, getter, setter)
end

--- Create a new service type
function service.create_service(client, service)
  local cls = class(ServiceBase)
  cls._client = client
  cls._name = service.name
  cls._invoke = bind1(client._invoke, client)

  -- Add class types to service
  for _,cls2 in ipairs(service.classes) do
    cls:_add_service_class(cls2)
  end

  -- Add procedures
  for _,procedure in ipairs(service.procedures) do
    if Attributes.is_a_procedure(procedure.attributes) then
      cls:_add_service_procedure(procedure)
    end
  end

  -- Add properties
  local properties = {}
  for _,procedure in ipairs(service.procedures) do
    if Attributes.is_a_property_accessor(procedure.attributes) then
      local name = Attributes.get_property_name(List(procedure.attributes))
      if not properties[name] then
        properties[name] = {}
      end
      if Attributes.is_a_property_getter(procedure.attributes) then
        properties[name]['get'] = procedure
      else
        properties[name]['set'] = procedure
      end
    end
  end
  for name, procedures in pairs(properties) do
    cls:_add_service_property(name, procedures['get'], procedures['set'])
  end

  return cls()
end

return service
